esp/platform/txsummary.cpp

```diff
@@ -20,22 +20,356 @@
+CTxSummary::TxEntryBase::TxEntryBase(const char* _key, const LogLevel _logLevel, const unsigned int _group, const char* _suffix): logLevel(_logLevel), group(_group), suffix(_suffix)
+{
+    // parse dot-delimited key
+    // 'name' is set to the rightmost element of _key
+    // 'fullname' is the entire _key
+
+    if(isEmptyString(_key))
+        name.set(_key);
+    else
+    {
+        const char* finger = strrchr(_key, '.');
+        if(finger)
+            name.set(++finger);
+        else
+            name.set(_key);
+    }
+
+    fullname.set(_key);
+}
+
+bool CTxSummary::TxEntryBase::shouldSerialize(const LogLevel requestedLevel, const unsigned int requestedGroup)
+{
+    if(logLevel > requestedLevel)
+        return false;
+
+    if(!(group & requestedGroup))
+        return false;
+
+    return true;
+}
+
+StringBuffer& CTxSummary::TxEntryBase::serialize(StringBuffer& buf, const LogLevel requestedLevel, const unsigned int requestedGroup, const unsigned int requestedStyle)
+{
+    return buf;
+}
+
+StringBuffer& CTxSummary::TxEntryStr::serialize(StringBuffer& buf, const LogLevel requestedLevel, const unsigned int requestedGroup, const unsigned int requestedStyle)
+{
+    if(!shouldSerialize(requestedLevel, requestedGroup))
+        return buf;
+
+    if(requestedStyle & TXSUMMARY_OUT_JSON)
+    {
+        // OK to return nullptr for value, then we serialize as JSON null
+        appendJSONStringValue(buf, name.str(), value.get(), false);
+    } else if (requestedStyle & TXSUMMARY_OUT_TEXT) {
+        buf.append(fullname);
+
+        if (value.length())
+            buf.appendf("=%s;", value.str());
+        else
+            buf.append(';');
+    }
+
+    return buf;
+}
+
+StringBuffer& CTxSummary::TxEntryBool::serialize(StringBuffer& buf, const LogLevel requestedLevel, const unsigned int requestedGroup, const unsigned int requestedStyle)
+{
+    if(!shouldSerialize(requestedLevel, requestedGroup))
+        return buf;
+
+    if(requestedStyle & TXSUMMARY_OUT_JSON)
+    {
+        appendJSONValue(buf, name.str(), value);
+    } else if (requestedStyle & TXSUMMARY_OUT_TEXT) {
+        buf.append(fullname);
+        buf.appendf("=%s;", boolToStr(value));
+    }
+
+    return buf;
+}
+
+
+StringBuffer& CTxSummary::TxEntryNumber::serialize(StringBuffer& buf, const LogLevel requestedLevel, const unsigned int requestedGroup, const unsigned int requestedStyle)
+{
+    if(!shouldSerialize(requestedLevel, requestedGroup))
+        return buf;
+
+    StringBuffer num;
+    appendValue(num);
+
+    if(requestedStyle & TXSUMMARY_OUT_JSON)
+    {
+        appendJSONStringValue(buf, name.str(), num.str(), true, false);
+    } else if (requestedStyle & TXSUMMARY_OUT_TEXT) {
+        buf.append(fullname);
+        buf.appendf("=%s%s;", num.str(), suffix.length() ? suffix.str() : "");
+    }
+
+    return buf;
+}
+
+bool CTxSummary::TxEntryTimer::shouldSerialize(const LogLevel requestedLevel, const unsigned int requestedGroup)
```

Comment:
Depending on resolution of timer initialization/update discussion, this variation may be unnecessary.

Reply:
I'm not understanding why. Don't we always want to appropriately filter out timers that don't match the requested level and/or group?

Reply:
If the timer's log level and group are immutable in an entry, the values can be shared with the entry itself (instead of ignoring them) and an override to get the values from the timer wouldn't be necessary.